File: HWRF/chem/KPP/kpp/kpp-2.1/drv/general_stochastic.c
int InitSaveData();
int SaveData();
int CloseSaveData();
int GenerateMatlab( char * prefix );
void GetMass( double CL[], double Mass[] );
void INTEGRATE( double TIN, double TOUT );
void Gillespie(int Nssa, double Volume, double* T, int NmlcV[], int NmlcF[]);
void Update_RCONST();

/* ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ */
int main()
/* ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ */
{
/*~~~> Output results file */
  FILE* fpDat;
/*~~~> No of molecules */
  int NmlcV[NVAR], NmlcF[NFIX];
/*~~~> No of reaction events per output step */
  int Nevents;
/*~~~> Local variables */
  int i;
  double T;

  Initialize();  
      
  fpDat = fopen("KPP_ROOT_stochastic.dat", "w");

  /* Translate initial values from conc. to molecules */
  /* Volume = 100.0; */
  Nevents = 20;
  for( i = 0; i < NVAR; i++ ) 
    NmlcV[i] = (int)(Volume*VAR[i]);
  for( i = 0; i < NFIX; i++ ) 
    NmlcF[i] = (int)(Volume*FIX[i]);
  
/*~~~> Begin Time Loop ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ */
  T = TSTART;
  while (T <= TEND) {
    printf("\n%6.1f%% %10.4f   ", (T-TSTART)/(TEND-TSTART)*100, T );
    for( i = 0; i < NVAR; i++ ) 
      printf( "%s=%d  ", SPC_NAMES[i], NmlcV[i] );

    fprintf(fpDat,"\n%g ", T );
    for( i = 0; i < NVAR; i++ ) 
        fprintf(fpDat,"%d  ", NmlcV[i]);
   
    Gillespie( Nevents, Volume, &T , NmlcV, NmlcF );
    
  }  /* while (T <= TEND) */  
/*~~~> End Time Loop ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ */

  fprintf(fpDat,"\n%g ", T );
  for( i = 0; i < NVAR; i++ ) 
      fprintf(fpDat,"%d  ", NmlcV[i]);
  fprintf(fpDat,"\n");
  fclose(fpDat);
  
  printf("\n");

  return 0;

}
/*~~~> End of MAIN function ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ */

File: chem/KPP/kpp/kpp-2.1/drv/general_stochastic.f90
PROGRAM KPP_ROOT_Driver

  USE KPP_ROOT_Model
  USE KPP_ROOT_Initialize, ONLY: Initialize

  IMPLICIT NONE
      
!~~~> No. of Stochastic events per simulation snaphot      
      INTEGER :: Nevents
!~~~> No. of Stochastic tau-steps simulation snaphot      
      INTEGER :: Nsteps
!~~~> No. of Molecules      
      INTEGER :: NmlcV(NVAR), NmlcF(NFIX)
!~~~> Random numbers     
      REAL :: r1, r2
!~~~> Local variables     
      INTEGER :: i
      KPP_REAL :: T, Tau, SCT(NREACT)

!~~~~> Initialize and prescribe volume and no. of events
      CALL Initialize()
!      Volume  = 1000.0d0
      Nevents = 5000
      Nsteps = 10
      Tau = (TEND-TSTART)/100.0
!~~~~~~~~~~~~~~~~     
      
!~~~> Translate initial values from conc. to NmlcVules
      NmlcV(1:NVAR) = INT(Volume*VAR(1:NVAR))
      NmlcF(1:NFIX) = INT(Volume*FIX(1:NFIX))
!~~~> Compute the stochastic reaction rates 
      CALL Update_RCONST()
      CALL StochasticRates( RCONST, Volume, SCT )   

!~~~> Save initial data
      OPEN(10, file='KPP_ROOT_stochastic.dat')
      WRITE(10,992) T, (NmlcV(i),i=1,NVAR)
            
!~~~> TIME loop starts
      T = TSTART
kron: DO WHILE (T < TEND)

        WRITE(6,991) T,(SPC_NAMES(i),NmlcV(i), i=1,NVAR)

!~~~> Choose here one of the following time-stepping routines
        CALL Gillespie(Nevents, T, SCT, NmlcV, NmlcF)
!        CALL TauLeap(Nsteps, Tau, T, SCT, NmlcV, NmlcF)
        
        WRITE(10,992) T, (NmlcV(i),i=1,NVAR)
        
      END DO kron
!~~~> TIME loop ends

      WRITE(6,991) T,(SPC_NAMES(i), NmlcV(i), i=1,NVAR)

      CLOSE(10)
      
991   FORMAT('T=',F12.3,20(A,'=',I5,'; '))
992   FORMAT(E24.14,200(1X,I12))


END PROGRAM KPP_ROOT_Driver

